Jetpack Partner Portal: Add actions to the WPCOM hosting license

yashwin opened the pull request:
Closes https://github.com/Automattic/jetpack-avalon/issues/44

## Proposed Changes

This PR implements adding actions(Setup site, Change domain, Edit site in WP Admin, Hosting configuration, Debug site & Revoke) to the Licenses table row when a WPCOM hosting license is displayed.

### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 31d88-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/actions-to-wpcom-hosting-license` and `yarn start-jetpack-cloud`.
2. Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
3. Visit the [Licenses](http://jetpack.cloud.localhost:3001/partner-portal/licenses) page.
4. Since we currently don't include WPCOM atomic sites, we must modify the code to test and verify these changes > Go to `/jetpack-cloud/sections/partner-portal/license-preview/index.tsx#L110` and replace the line with:

```
const isAtomic = !! blogId;
```

5. Verify that you can see the `...` actions icon instead of the down arrow icon. > Click the `...` icon and verify you can see the actions as shown below. > Verify all the actions work as expected.
6. Verify the actions look good on small-screen devices
7. Verify that events are tracked when any action is clicked. 

<img width="244" alt="Screenshot 2023-09-04 at 1 01 54 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/5aaae564-86ea-4d93-a698-fd8160ce44c9">


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yashwin committed:
Create a hook that returns the licenses

yashwin committed:
Create component to render the license actions

yashwin committed:
Render the actions for atomic sites

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/license-preview/license-actions.tsx:
```diff
@@ -0,0 +1,75 @@
+	const buttonActionRef = useRef< HTMLButtonElement | null >( null );
+
+	const [ isOpen, setIsOpen ] = useState( false );
+	const [ openModal, setOpenModal ] = useState< string | null >( null );
```
Non-blocking:

I'm a little confused here, as to the intent behind making `openModal` a string. It appears that we only ever care if a value is present or not present, and only in the case of the Revoke action being clicked.

Do we have concrete plans to add more dialogs for other action types? If not, I wonder if it would be slightly clearer and more resilient to convert this to something like

```ts
const [ showRevokeDialog, setShowRevokeDialog ] = useState< boolean >( false );

// handleActionClick
if ( action.type === 'revoke' ) {
	setShowRevokeDialog( true );
}

// useLicenseActions
{
	type: 'setup',
	...
},
{
	type: 'revoke',
	...
}
```

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/types.ts:
```diff
@@ -34,3 +34,13 @@ export interface AssignLicenceProps {
 	suggestedProduct?: string;
 }
+
+export interface LicenceAction {
```
Non-critical:

It's probably a good idea for us to settle on one uniform spelling of license/licence. I know both are valid depending on the dialect of English being used, but I see both spellings in this PR.

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/license-preview/use-license-actions.ts:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	const siteSlug = urlToSlug( siteUrl );
+	const debugUrl = `https://jptools.wordpress.com/debug/?url=${ siteUrl }`;
```
Non-blocking:

Any reason this link href has its own variable, while all the others are inline? I see it's referenced in a boolean condition, but I'm not sure it's necessary (see my comment on line 68).

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/license-preview/use-license-actions.ts:
```diff
@@ -0,0 +1,78 @@
+			href: debugUrl,
+			onClick: () => handleClickMenuItem( 'calypso_jetpack_licenses_debug_site_click' ),
+			isExternalLink: true,
+			isEnabled: licenseState === LicenseState.Attached && !! debugUrl,
```
`debugUrl` is always truthy, isn't it?

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/license-preview/use-license-actions.ts:
```diff
@@ -0,0 +1,78 @@
+
+	const licenseState = getLicenseState( attachedAt, revokedAt );
+
+	return [
```
Non-blocking:

This return value can be memoized and written to have almost all of its dependencies self-contained; e.g.,

```ts
return useMemo( () => {
	const handleClickMenuItem = /* ... */;
	const licenseState = /* ... */;
	const siteSlug = /* ... */;

	return [
		// { ... }, { ... }
	];
}, [ translate, /* ... */ ] );
```

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/license-preview/index.tsx:
```diff
@@ -56,6 +60,8 @@ export default function LicensePreview( {
 	const licenseState = getLicenseState( attachedAt, revokedAt );
 	const domain = siteUrl ? getUrlParts( siteUrl ).hostname || siteUrl : '';
 
+	const site = useSelector( ( state ) => ( blogId ? getSite( state, blogId ) : null ) );
```
Non-blocking:

The `getSite` selector is already written to handle the null case. Then again, I appreciate that this could just be defensive programming. 🙂

mattgawarecki commented on client/jetpack-cloud/sections/partner-portal/license-preview/index.tsx:
```diff
@@ -101,6 +107,9 @@ export default function LicensePreview( {
 		}
 	}, [] );
 
+	const isAtomic = site?.is_wpcom_atomic;
+	const isSiteAtomic = isEnabled( 'jetpack/pro-dashboard-wpcom-atomic-hosting' ) && isAtomic;
```
Non-blocking:

If we're only using `isAtomic` in one place, I think using its source value directly inline still provides a reasonable level of readability:

```suggestion
	const isSiteAtomic = isEnabled( 'jetpack/pro-dashboard-wpcom-atomic-hosting' ) && site?.is_wpcom_atomic;
```

mattgawarecki reviewed (approved):
I tested functionally and everything seems to be correct. I also took a relatively close look into the code and found a few areas of possible improvement, but nothing terribly disruptive or risky. Feel free to address my comments (or not) at your own judgment.

yashwin committed:
Address PR comments

yashwin commented:
@mattgawarecki: Thanks for your very thorough review! 

I have addressed all your comments by making the changes: https://github.com/Automattic/wp-calypso/pull/81345/commits/a5ae15cf2b87d7f5e54160ab762aefa1fe39986b

mattgawarecki commented:
@yashwin I've reviewed your changes and my approval still stands. Great work 🙂

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8870584

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.
